File: src/lib/regions.ts
import { Region } from "@/generated/prisma/enums";

export const REGIONS = [
  "CORNWALL",
  "DEVON",
  "SOMERSET",
  "BRISTOL",
  "GLOUCESTERSHIRE",
  "WILTSHIRE",
  "DORSET",
  "HAMPSHIRE",
  "WEST_SUSSEX",
  "SURREY",
  "BERKSHIRE",
  "GREATER_LONDON",
  "OXFORDSHIRE",
] as const satisfies readonly Region[];

export type RegionKey = (typeof REGIONS)[number];

export interface RegionMeta {
  key: RegionKey;
  label: string;
  /// Approximate administrative centre — used to place a marker when we only
  /// know the county, not a street address.
  centroid: { lat: number; lng: number };
  /// The 2013 local-authority districts that make up this ceremonial county.
  /// Used by scripts/build-geo.ts to assemble the map, and to resolve a
  /// Companies House registered-office district into one of our regions.
  districts: string[];
}

export const REGION_META: Record<RegionKey, RegionMeta> = {
  CORNWALL: {
    key: "CORNWALL",
    label: "Cornwall",
    centroid: { lat: 50.4, lng: -4.8 },
    districts: ["Cornwall", "Isles of Scilly"],
  },
  DEVON: {
    key: "DEVON",
    label: "Devon",
    centroid: { lat: 50.72, lng: -3.7 },
    districts: [
      "East Devon",
      "Exeter",
      "Mid Devon",
      "North Devon",
      "Plymouth",
      "South Hams",
      "Teignbridge",
      "Torbay",
      "Torridge",
      "West Devon",
    ],
  },
  SOMERSET: {
    key: "SOMERSET",
    label: "Somerset",
    centroid: { lat: 51.13, lng: -2.85 },
    districts: [
      "Bath and North East Somerset",
      "Mendip",
      "North Somerset",
      "Sedgemoor",
      "South Somerset",
      "Taunton Deane",
      "West Somerset",
    ],
  },
  BRISTOL: {
    key: "BRISTOL",
    label: "Bristol",
    centroid: { lat: 51.4545, lng: -2.5879 },
    districts: ["Bristol, City of"],
  },
  GLOUCESTERSHIRE: {
    key: "GLOUCESTERSHIRE",
    label: "Gloucestershire",
    centroid: { lat: 51.83, lng: -2.2 },
    districts: [
      "Cheltenham",
      "Cotswold",
      "Forest of Dean",
      "Gloucester",
      "South Gloucestershire",
      "Stroud",
      "Tewkesbury",
    ],
  },
  WILTSHIRE: {
    key: "WILTSHIRE",
    label: "Wiltshire",
    centroid: { lat: 51.35, lng: -1.95 },
    districts: ["Swindon", "Wiltshire"],
  },
  DORSET: {
    key: "DORSET",
    label: "Dorset",
    centroid: { lat: 50.75, lng: -2.35 },
    districts: [
      "Bournemouth",
      "Christchurch",
      "East Dorset",
      "North Dorset",
      "Poole",
      "Purbeck",
      "West Dorset",
      "Weymouth and Portland",
    ],
  },
  HAMPSHIRE: {
    key: "HAMPSHIRE",
    label: "Hampshire",
    centroid: { lat: 51.05, lng: -1.31 },
    districts: [
      "Basingstoke and Deane",
      "East Hampshire",
      "Eastleigh",
      "Fareham",
      "Gosport",
      "Hart",
      "Havant",
      "New Forest",
      "Portsmouth",
      "Rushmoor",
      "Southampton",
      "Test Valley",
      "Winchester",
    ],
  },
  WEST_SUSSEX: {
    key: "WEST_SUSSEX",
    label: "West Sussex",
    centroid: { lat: 50.94, lng: -0.47 },
    districts: [
      "Adur",
      "Arun",
      "Chichester",
      "Crawley",
      "Horsham",
      "Mid Sussex",
      "Worthing",
    ],
  },
  SURREY: {
    key: "SURREY",
    label: "Surrey",
    centroid: { lat: 51.27, lng: -0.44 },
    districts: [
      "Elmbridge",
      "Epsom and Ewell",
      "Guildford",
      "Mole Valley",
      "Reigate and Banstead",
      "Runnymede",
      "Spelthorne",
      "Surrey Heath",
      "Tandridge",
      "Waverley",
      "Woking",
    ],
  },
  BERKSHIRE: {
    key: "BERKSHIRE",
    label: "Berkshire",
    centroid: { lat: 51.45, lng: -0.98 },
    districts: [
      "Bracknell Forest",
      "Reading",
      "Slough",
      "West Berkshire",
      "Windsor and Maidenhead",
      "Wokingham",
    ],
  },
  GREATER_LONDON: {
    key: "GREATER_LONDON",
    label: "Greater London",
    centroid: { lat: 51.5074, lng: -0.1278 },
    // Resolved by ONS code prefix E09 in scripts/build-geo.ts.
    districts: [],
  },
  OXFORDSHIRE: {
    key: "OXFORDSHIRE",
    label: "Oxfordshire",
    centroid: { lat: 51.77, lng: -1.28 },
    districts: [
      "Cherwell",
      "Oxford",
      "South Oxfordshire",
      "Vale of White Horse",
      "West Oxfordshire",
    ],
  },
};

export const REGION_LABELS: Record<RegionKey, string> = Object.fromEntries(
  REGIONS.map((r) => [r, REGION_META[r].label]),
) as Record<RegionKey, string>;

/** Reverse lookup: district / town name -> region. Built once at module load. */
const DISTRICT_INDEX: Map<string, RegionKey> = (() => {
  const m = new Map<string, RegionKey>();
  for (const r of REGIONS) {
    for (const d of REGION_META[r].districts) m.set(d.toLowerCase(), r);
    m.set(REGION_META[r].label.toLowerCase(), r);
  }
  return m;
})();

/**
 * Best-effort resolution of a free-text location (typically a Companies House
 * registered-office locality or county line) to one of the 13 in-scope regions.
 * Returns null when the address is out of scope — callers must treat null as
 * "do not track", not as "default to London".
 */
export function resolveRegion(input: string | null | undefined): RegionKey | null {
  if (!input) return null;
  const text = input.toLowerCase();

  // Exact district match first (most specific).
  const direct = DISTRICT_INDEX.get(text.trim());
  if (direct) return direct;

  // London postcode districts and the common "London" locality.
  if (/\blondon\b/.test(text) && !/\blondonderry\b/.test(text)) return "GREATER_LONDON";

  for (const [name, region] of DISTRICT_INDEX) {
    // Word-boundary match so "Reading" doesn't fire on "Spreading".
    const re = new RegExp(`\\b${name.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}\\b`);
    if (re.test(text)) return region;
  }

  // A handful of high-frequency towns that aren't district names.
  const towns: Array<[RegExp, RegionKey]> = [
    [/\b(truro|falmouth|newquay|penzance|st ives|bodmin|redruth)\b/, "CORNWALL"],
    [/\b(barnstaple|tiverton|newton abbot|paignton|totnes|okehampton)\b/, "DEVON"],
    [/\b(yeovil|bridgwater|frome|wells|glastonbury|bath|weston-super-mare|clevedon)\b/, "SOMERSET"],
    [/\b(cirencester|tetbury|dursley|thornbury|yate|nailsworth)\b/, "GLOUCESTERSHIRE"],
    [/\b(salisbury|chippenham|trowbridge|devizes|marlborough|malmesbury)\b/, "WILTSHIRE"],
    [/\b(dorchester|bridport|sherborne|wareham|blandford|swanage|wimborne)\b/, "DORSET"],
    [/\b(andover|romsey|petersfield|alton|farnborough|aldershot|lymington|hedge end)\b/, "HAMPSHIRE"],
    [/\b(haywards heath|burgess hill|east grinstead|bognor|littlehampton|shoreham)\b/, "WEST_SUSSEX"],
    [/\b(esher|weybridge|cobham|leatherhead|dorking|farnham|godalming|camberley|staines|redhill|oxted)\b/, "SURREY"],
    [/\b(newbury|maidenhead|ascot|thatcham|sandhurst|hungerford|eton)\b/, "BERKSHIRE"],
    [/\b(banbury|bicester|abingdon|witney|didcot|henley-on-thames|thame|wallingford)\b/, "OXFORDSHIRE"],
  ];
  for (const [re, region] of towns) if (re.test(text)) return region;

  return null;
}

export function isInScope(region: string | null | undefined): region is RegionKey {
  return !!region && (REGIONS as readonly string[]).includes(region);
}

File: wealthscan/regions.py
"""The 13 in-scope counties, and resolving free text to one of them.

Geographic scope is enforced in code, not by convention: a record whose
location cannot be resolved is discarded rather than defaulted, so nothing
outside the advisor's patch ever enters the pipeline.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

REGIONS: tuple[str, ...] = (
    "Cornwall",
    "Devon",
    "Somerset",
    "Bristol",
    "Gloucestershire",
    "Wiltshire",
    "Dorset",
    "Hampshire",
    "West Sussex",
    "Surrey",
    "Berkshire",
    "Greater London",
    "Oxfordshire",
)


@dataclass(frozen=True)
class Region:
    name: str
    lat: float
    lon: float
    #: Districts and major towns, used both for matching and for search queries.
    places: tuple[str, ...]


REGION_DATA: dict[str, Region] = {
    "Cornwall": Region("Cornwall", 50.40, -4.80, (
        "Cornwall", "Truro", "Falmouth", "Newquay", "Penzance", "St Ives",
        "Bodmin", "Redruth", "Camborne", "Helston", "Launceston", "Bude",
        "St Austell", "Wadebridge", "Padstow", "Isles of Scilly",
    )),
    "Devon": Region("Devon", 50.72, -3.70, (
        "Devon", "Exeter", "Plymouth", "Torbay", "Torquay", "Paignton",
        "Barnstaple", "Newton Abbot", "Exmouth", "Tiverton", "Totnes",
        "Okehampton", "Bideford", "Dartmouth", "Ilfracombe", "Honiton",
        "Sidmouth", "South Hams", "Teignbridge", "Torridge",
    )),
    "Somerset": Region("Somerset", 51.13, -2.85, (
        "Somerset", "Taunton", "Yeovil", "Bath", "Bridgwater", "Wells",
        "Frome", "Glastonbury", "Weston-super-Mare", "Clevedon", "Portishead",
        "Street", "Shepton Mallet", "Minehead", "Mendip", "Sedgemoor",
    )),
    "Bristol": Region("Bristol", 51.4545, -2.5879, (
        "Bristol", "Clifton", "Bedminster", "Avonmouth", "Redcliffe",
    )),
    "Gloucestershire": Region("Gloucestershire", 51.83, -2.20, (
        "Gloucestershire", "Gloucester", "Cheltenham", "Cirencester", "Stroud",
        "Tewkesbury", "Cotswold", "Cotswolds", "Forest of Dean", "Tetbury",
        "Dursley", "Thornbury", "Yate", "Nailsworth", "South Gloucestershire",
    )),
    "Wiltshire": Region("Wiltshire", 51.35, -1.95, (
        "Wiltshire", "Swindon", "Salisbury", "Chippenham", "Trowbridge",
        "Devizes", "Marlborough", "Malmesbury", "Warminster", "Melksham",
    )),
    "Dorset": Region("Dorset", 50.75, -2.35, (
        "Dorset", "Bournemouth", "Poole", "Christchurch", "Dorchester",
        "Weymouth", "Bridport", "Sherborne", "Wareham", "Blandford",
        "Swanage", "Wimborne", "Sandbanks", "Purbeck", "Ferndown",
    )),
    "Hampshire": Region("Hampshire", 51.05, -1.31, (
        "Hampshire", "Southampton", "Portsmouth", "Winchester", "Basingstoke",
        "Andover", "Eastleigh", "Fareham", "Gosport", "Havant", "Aldershot",
        "Farnborough", "Romsey", "Petersfield", "Alton", "Lymington",
        "New Forest", "Test Valley", "Hart", "Rushmoor", "Whiteley",
    )),
    "West Sussex": Region("West Sussex", 50.94, -0.47, (
        "West Sussex", "Chichester", "Crawley", "Horsham", "Worthing",
        "Bognor Regis", "Littlehampton", "Shoreham", "Haywards Heath",
        "Burgess Hill", "East Grinstead", "Arun", "Adur", "Mid Sussex",
    )),
    "Surrey": Region("Surrey", 51.27, -0.44, (
        "Surrey", "Guildford", "Woking", "Epsom", "Esher", "Weybridge",
        "Cobham", "Leatherhead", "Dorking", "Farnham", "Godalming",
        "Camberley", "Staines", "Redhill", "Reigate", "Oxted", "Elmbridge",
        "Mole Valley", "Runnymede", "Waverley", "Tandridge",
    )),
    "Berkshire": Region("Berkshire", 51.45, -0.98, (
        "Berkshire", "Reading", "Newbury", "Maidenhead", "Slough", "Bracknell",
        "Windsor", "Wokingham", "Ascot", "Thatcham", "Sandhurst", "Hungerford",
        "Eton", "West Berkshire",
    )),
    "Greater London": Region("Greater London", 51.5074, -0.1278, (
        "London", "Greater London", "City of London", "Westminster", "Camden",
        "Islington", "Hackney", "Shoreditch", "Southwark", "Lambeth",
        "Wandsworth", "Kensington", "Chelsea", "Fulham", "Hammersmith",
        "Richmond upon Thames", "Kingston upon Thames", "Croydon", "Bromley",
        "Barnet", "Brent", "Ealing", "Enfield", "Greenwich", "Haringey",
        "Harrow", "Havering", "Hillingdon", "Hounslow", "Lewisham", "Merton",
        "Newham", "Redbridge", "Sutton", "Tower Hamlets", "Waltham Forest",
        "Bexley", "Barking and Dagenham", "Mayfair", "Canary Wharf",
    )),
    "Oxfordshire": Region("Oxfordshire", 51.77, -1.28, (
        "Oxfordshire", "Oxford", "Banbury", "Bicester", "Abingdon", "Witney",
        "Didcot", "Henley-on-Thames", "Thame", "Wallingford", "Cherwell",
        "Vale of White Horse", "Wantage", "Chipping Norton", "Harwell",
    )),
}

#: Ambiguous place names that appear in unrelated contexts and would otherwise
#: create false positives. Matched only with an explicit county nearby.
AMBIGUOUS_PLACES = frozenset({
    "Bath", "Reading", "Windsor", "Richmond upon Thames", "Chelsea",
    "Kensington", "Hart", "Street", "Wells", "Sutton", "Cornwall",
})

#: Terms that look like an in-scope place but are not.
NEGATIVE_TERMS = (
    "New London", "London Ontario", "Londonderry", "Bristol Connecticut",
    "Bristol Tennessee", "Bristol Virginia", "Reading Pennsylvania",
    "Chelsea FC", "Chelsea Football", "Hampshire College",
    "New Hampshire", "Windsor Ontario", "Richmond Virginia",
)


def _compile(places: tuple[str, ...]) -> re.Pattern[str]:
    alternatives = sorted((re.escape(p) for p in places), key=len, reverse=True)
    return re.compile(r"\b(" + "|".join(alternatives) + r")\b", re.IGNORECASE)


_PATTERNS: dict[str, re.Pattern[str]] = {
    name: _compile(region.places) for name, region in REGION_DATA.items()
}
_AMBIGUOUS_LOWER = {p.lower() for p in AMBIGUOUS_PLACES}


def resolve_region(text: str | None) -> tuple[str | None, str | None]:
    """Resolve free text to one of the 13 counties.

    Returns ``(region, matched_place)``. ``(None, None)`` means out of scope,
    which callers must treat as "discard", never as "assume London".

    An ambiguous place name on its own is not enough: "Bath" appears in far more
    articles about bathrooms than about Somerset, so a bare ambiguous match is
    only accepted when the county name also appears.
    """
    if not text:
        return None, None

    for negative in NEGATIVE_TERMS:
        if negative.lower() in text.lower():
            text = re.sub(re.escape(negative), " ", text, flags=re.IGNORECASE)

    best: tuple[str, str] | None = None
    for name, pattern in _PATTERNS.items():
        match = pattern.search(text)
        if not match:
            continue
        matched = match.group(1)

        if matched.lower() in _AMBIGUOUS_LOWER and matched.lower() != name.lower():
            # Require corroboration from the county name itself.
            if not re.search(r"\b" + re.escape(name) + r"\b", text, re.IGNORECASE):
                continue

        # An exact county-name match beats a town match.
        if matched.lower() == name.lower():
            return name, matched
        if best is None:
            best = (name, matched)

    return best if best else (None, None)


def region_centroid(name: str) -> tuple[float, float]:
    region = REGION_DATA[name]
    return region.lat, region.lon
